certora/specs/TimelockController.spec

```
@@ -128,151 +166,91 @@ rule minDelayOnlyChange(env e) {
 } {
-    bool isExecutorOrOpen = hasRole(EXECUTOR_ROLE(), e.msg.sender) || hasRole(EXECUTOR_ROLE(), 0);
-
-    calldataarg args;
-    f@withrevert(e, args);
-
-    assert !isExecutorOrOpen => lastReverted, "Only executor can execute";
+    bytes32 otherId; uint256 otherTimestamp = getTimestamp(otherId);
+
+    uint8 stateBefore            = state(id);
+    bool  isOperationReadyBefore = isOperationReady(e, id);
+    bool  isExecutorOrOpen       = hasRole(EXECUTOR_ROLE(), e.msg.sender) || hasRole(EXECUTOR_ROLE(), 0);
+    bool  predecessorDependency  = predecessor == 0 || isDone(predecessor);
+
+    helperExecuteWithRevert(e, f, id, predecessor);
+
+    // The underlying transaction can revert, and that would cause the execution to revert.
+    //
+    // We can check that all non reverting calls meet the requirements in terms of proposal readiness, access control
+    // and predecessor dependency. We can't however guarantee that these requirements being meet ensure liveness of the
+    // proposal, because the proposal can revert for reasons beyond our control.
+    if (lastReverted) {
+        // assert stateBefore != PENDING() || !isOperationReadyBefore || !predecessorDependency || !isExecutorOrOpen, "Unexpected revert";
+        assert true;
+    } else {
+        assert stateBefore == PENDING() && state(id) == DONE(), "State transition violation";
+        assert isExecutorOrOpen, "Unauthorized execute";
+        assert isOperationReadyBefore, "Execute before ready";
+        assert predecessorDependency, "Predecessor dependency violation";
+        assert otherTimestamp != getTimestamp(otherId) => id == otherId, "Other proposal affected";
+    }
 }
 
 /*
 ┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
-│ Access control: only users with the CANCELLER_ROLE can call cancel                                                  │
+│ Rule: cancel liveliness and effets                                                                                  │
 └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
 */
-rule onlyCanceler(env e, bytes32 id, method f) filtered { f ->
-    f.selector == cancel(bytes32).selector
-} {
-    bool isCanceler = hasRole(CANCELLER_ROLE(), e.msg.sender);
+rule cancel(env e, bytes32 id) {
+    require nonpayable(e);
+
+    bytes32 otherId; uint256 otherTimestamp = getTimestamp(otherId);
+
+    uint8 stateBefore = state(id);
+    bool  isCanceller = hasRole(CANCELLER_ROLE(), e.msg.sender);
 
-    calldataarg args;
-    f@withrevert(e, args);
+    cancel@withrevert(e, id);
 
-    assert !isCanceler => lastReverted, "Only canceler can cancel";
+    if (lastReverted) {
+        assert stateBefore != PENDING() || !isCanceller, "Unexpected revert";
+    } else {
+        assert stateBefore == PENDING() && state(id) == UNSET(), "State transition violation";
```

Same question/suggestion as above.

```suggestion
    bool success = !lastReverted;

    assert success <=> (stateBefore == PENDING() && isCanceller), "Unexpected revert";

    if (success) {
```